src/LuzFaltex.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs

```diff
@@ -0,0 +1,697 @@
+        }
+
+        return null;
+    }
+
+    private ServiceCallSite? TryCreateOpenGeneric(Type serviceType, CallSiteChain callSiteChain)
+    {
+        if (serviceType.IsConstructedGenericType
+            && _descriptorLookup.TryGetValue(serviceType.GetGenericTypeDefinition(), out var descriptor))
+        {
+            return TryCreateOpenGeneric(descriptor.Last, serviceType, callSiteChain, DefaultSlot, true);
+        }
+
+        return null;
+    }
+
+    private ServiceCallSite? TryCreateEnumerable(Type serviceType, CallSiteChain callSiteChain)
+    {
+        var callSiteKey = new ServiceCacheKey(serviceType, DefaultSlot);
+        if (_callSiteCache.TryGetValue(callSiteKey, out var serviceCallSite))
+        {
+            return serviceCallSite;
+        }
+
+        try
+        {
+            callSiteChain.Add(serviceType);
+
+            if (serviceType.IsConstructedGenericType &&
+                serviceType.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+            {
+                var itemType = serviceType.GenericTypeArguments[0];
+                var cacheLocation = CallSiteResultCacheLocation.Root;
+
+                var callSites = new List<ServiceCallSite>();
+
+                // If item type is not generic we can safely use descriptor cache
+                if (!itemType.IsConstructedGenericType &&
+                    _descriptorLookup.TryGetValue(itemType, out var descriptors))
+                {
+                    for (var i = 0; i < descriptors.Count; i++)
+                    {
+                        var descriptor = descriptors[i];
+
+                        // Last service should get slot 0
+                        var slot = descriptors.Count - i - 1;
+                        // There may not be any open generics here
+                        var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot);
+                        Debug.Assert(callSite != null);
+
+                        cacheLocation = GetCommonCacheLocation(cacheLocation, callSite.Cache.Location);
+                        callSites.Add(callSite);
+                    }
+                }
+                else
+                {
+                    var slot = 0;
+                    var serviceCount = _serviceCollections.CountServices();
+                    for (int i = serviceCount.Count - 1; i >= 0; i--)
+                    {
+                        var descriptor = serviceCount[i];
+                        var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot) ??
+                                       TryCreateOpenGeneric(descriptor, itemType, callSiteChain, slot, false);
+
+                        if (callSite != null)
+                        {
+                            slot++;
+
+                            cacheLocation = GetCommonCacheLocation(cacheLocation, callSite.Cache.Location);
+                            callSites.Add(callSite);
+                        }
+                    }

```

Comment:
It seems that every time a service collection is added, that then new callsites are created (even ones that already existed) which results in new instances being created. Oops.